### DISCUSSION TITLE
LandCover.ai as a Geospatial Dataset

### DISCUSSION THREAD

**Discussion by adrianboguszewski** [torchgeo/torchgeo#1104] [General]
Hi. Thanks for incorporating LandCover.ai into torchgeo! I don't know if you're aware that LandCover.ai consist of GeoTiffs so that it could be treated as a Geospatial Dataset. Of course, that information is lost when using the split.py script to generate tiles for training. The script was provided to ensure reproducibility and comparability to the paper results. However, the dataset itself was designed as 41 GeoTiffs + masks :)

**Comment by adamjstewart** [comment]
Hi @adrianboguszewski, yes we did notice that! Our first implementation was designed for the reason you mention, for benchmarking and comparison against other papers. However, we would love to add a second implementation for a GeoDataset. Would you be interested in submitting a PR for this? [This tutorial](https://torchgeo.readthedocs.io/en/stable/tutorials/custom_raster_dataset.html) gives a good explanation of how to create a new GeoDataset, and the tests should be easy to copy from other datasets.

**Reply by adrianboguszewski** [reply]
Why not. When I find time, I can submit a PR :)

**Reply by adrianboguszewski** [reply]
If LandCover.ai contains images and masks (tif and tif), should I use RasterDataset as a base class? It seems to support either images or masks.

**Reply by adamjstewart** [reply]
Ah right, that was one of the remaining TODOs in figuring out how the API should work.

There are a few options. You could just subclass GeoDataset directly, but you would end up duplicating a lot of the code in RasterDataset, which is unideal (ChesapeakeCVPR does this). Another idea would be to subclass RasterDataset but override `__getitem__` so that it loads both the image and mask. A third option would be to create two RasterDatasets and then create an IntersectionDataset like:
```python
class _LandCoverAIImage(RasterDataset):
    pass

class _LandCoverAIMask(RasterDataset):
    pass

class LandCoverAIGeo(IntersectionDataset):  # name TBD
    def __init__(...):
        image_ds = _LandCoverAIImage(...)
        mask_ds = _LandCoverAIMask(...)
        super().__init__(image_ds, mask_ds)
```
Fourth option is to change the API of RasterDataset to be able to handle both images and masks but that requires some thought.

Out of curiosity, are all of the images in LandCover.ai in the same CRS? Do they have any overlap or shared borders?

**Reply by adrianboguszewski** [reply]
Yes, the CRS is the same. Most of the images don't share borders. However, there are a few that overlap, e.g. 
![image](https://user-images.githubusercontent.com/4547501/219650952-a3352e0a-af5c-418b-9dd1-71baf375e3c6.png)

Taking into account, you want to update RasterDataset API in the future, I think the best solution is the third option?

**Reply by adamjstewart** [reply]
Yes, 2nd or 3rd option is probably best at the moment.

**Reply by adrianboguszewski** [reply]
Oh, I meant 2nd option, not 3rd. My mistake. But it looks like we agree :)
